src/picongpu/include/plugins/hdf5/openPMD/patchReader.hpp

```diff
@@ -0,0 +1,115 @@
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "plugins/common/particlePatches.hpp"
+
+#if( ENABLE_HDF5 == 1 )
+#  include <splash/splash.h>
+#endif
+
+#include <vector>
+#include <list>
+#include <string>
+#include <iostream>
+#include <typeinfo>
+
+namespace picongpu
+{
+namespace hdf5
+{
+namespace openPMD
+{
+    class PatchReader;
+
+#if( ENABLE_HDF5 == 1 )
+    /** Functor to populate and validate the list of particle patches
+     */
+    class PatchReader
+    {
+    private:
+        /** Determine the variable type for `offset` and `extent`
+         *
+         * In particle patches, the `offset` and `extent` can be of
+         * user-defined types. This function allows to determine which
+         * one was used and how many patches exist.
+         *
+         * @note currently we force the type to be `uint64_t`,
+         *       we can implement type conversions later on
+         * @note currently we force the number of patches
+         *       to stay constant during restarts
+         *
+         * @see operator() for parameters
+         */
+        void checkSpatialTypeSize(
+            splash::DataCollector* const dc,
+            const uint32_t availableRanks,
+            const int32_t id,
+            const std::string particlePatchPathComponent
+        ) const;
+
+        /** Read a specific record component of the particle patch
+         *
+         * Read for example: numParticles or offset/x
+         *
+         * @param[in]  dc pointer to an open splash::DataCollector
+         * @param[in]  availableRanks MPI ranks in the restarted simulation
+         *             that are currently waiting to find patches
+         * @param[in]  id time step to read
+         * @param[in]  particlePatchPathComponent string such as
+         *             "particles/e/particlePatches/numParticles" or
+         *             "particles/e/particlePatches/offset/x"
+         * @param[out] dest beginning of c-array of length size()
+         *             to write the patch record component to
+         */
+        void readPatchAttribute(
+            splash::DataCollector* const dc,
+            const uint32_t availableRanks,
+            const int32_t id,
+            const std::string particlePatchPathComponent,
+            uint64_t* const dest
```

Comment:
the idea is: if someone needs other offsets already -> tiny pre-process script in python.

later: clean conversions in c++ on our side (too much boilerplate for now).